sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/models/ProgressReporter.java

```diff
@@ -0,0 +1,174 @@
+            this.blockProgress += bytesTransferred;
+        }
+
+        void rewindProgress() {
+            this.blockProgress = 0;
+        }
+
+        Flux<ByteBuffer> addProgressReporting(Flux<ByteBuffer> data) {
+            return Mono.just(this).flatMapMany(progressReporter -> {
+                /*
+                Each time there is a new subscription, we will rewind the progress. This is desirable specifically
+                for retries, which resubscribe on each try. The first time this Flux is subscribed to, the
+                rewind will be a noop as there will have been no progress made. Subsequent rewinds will work as
+                expected.
+                 */
+                progressReporter.rewindProgress();
+
+                /*
+                Every time we emit some data, report it to the Tracker, which will pass it on to the end user.
+                 */
+                return data.doOnNext(buffer -> progressReporter.reportProgress(buffer.remaining()));
+            });
+        }
+    }
+
+    /**
+     * This type is used to keep track of the total amount of data transferred for a single request. This is the type we
+     * will use when the customer uses the factory to add progress reporting to their Flowable. We need this additional
+     * type because we can't keep local state directly as lambdas require captured local variables to be effectively
+     * final.
+     */
+    private static class SequentialProgressReporter extends ProgressReporterImpl {
+        SequentialProgressReporter(ProgressReceiver progressReceiver) {
+            super(progressReceiver);
+        }
+
+        @Override
+        public void reportProgress(long bytesTransferred) {
+            super.reportProgress(bytesTransferred);
+            this.progressReceiver.reportProgress(this.blockProgress);
+        }
+    }
+
+    /**
+     * This type is used to keep track of the total amount of data transferred as a part of a parallel upload in order
+     * to coordinate progress reporting to the end user. We need this additional type because we can't keep local state
+     * directly as lambdas require captured local variables to be effectively final.
+     */
+    private static class ParallelProgressReporter extends ProgressReporterImpl {
+        /*
+        This lock will be instantiated by the operation initiating the whole transfer to coordinate each
+        ProgressReporterImpl.
+         */
+        private final Lock transferLock;
+
+        /*
+        We need an AtomicLong to be able to update the value referenced. Because we are already synchronizing with the
+        lock, we don't incur any additional performance hit here by the synchronization.
+         */
+        private final AtomicLong totalProgress;
+
+        ParallelProgressReporter(ProgressReceiver progressReceiver, AtomicLong totalProgress) {
+            super(progressReceiver);
+            this.transferLock = new ReentrantLock();
```

Comment:
Each instance of ParallelProgressReporter has its own lock. That probably wouldn't do you any good. Where do you see race condition?

Reply:
To be honest, this is code extracted from Storage SDK. If we don't see any benefit here, we can just get rid of it.